### PR TITLE
Document unsafety in core::{panicking, alloc::layout, hint, iter::adapters::zip}

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
Helps with #66219.
r? @Mark-Simulacrum do you want to continue reading safety comments? :D

### CODE REVIEW

File: src/libcore/alloc/layout.rs [hunk 1]
```diff
@@ -143,7 +143,7 @@ impl Layout {
     #[unstable(feature = "alloc_layout_extra", issue = "55724")]
     #[inline]
     pub const fn dangling(&self) -> NonNull<u8> {
-        // align is non-zero and a power of two
+        // SAFETY: align is non-zero and a power of two
```
Comment (Mark-Simulacrum) [inline src/libcore/alloc/layout.rs:146]: Alignment being a power of two seems... irrelevant here?

File: src/libcore/hint.rs [hunk 2]
```diff
@@ -81,10 +81,13 @@ pub fn spin_loop() {
     {
         #[cfg(target_arch = "aarch64")]
         {
+            // SAFETY: the `cfg` attr ensures that we only execute this on aarch64 targets.
```
Comment (Mark-Simulacrum) [inline src/libcore/hint.rs:84]: Why do we have this setup so strangely? Can we "unwrap" these inner cfg's?

AFAICT we should also be able to do the same for the x86/x86_64 cfg's, just gating on target_feature = "sse2" in the wrapper (or inlining that).

File: src/libcore/iter/adapters/zip.rs [hunk 3]
```diff
@@ -203,11 +203,15 @@ where
             let i = self.index;
             self.index += 1;
             if A::may_have_side_effect() {
+                // SAFETY: rhe usage of `cmp::min` to calculate `delta`
```
Comment (Mark-Simulacrum) [inline src/libcore/iter/adapters/zip.rs:206]:
```suggestion
                // SAFETY: the usage of `cmp::min` to calculate `delta`
```

File: src/libcore/hint.rs (hunk 2 above)
Comment (LeSeulArtichaut) [reply src/libcore/hint.rs:84]: This should probably be done in a separate PR?

**Comment by LeSeulArtichaut** [comment]
@Mark-Simulacrum Resolved your review comments!

File: src/libcore/hint.rs (hunk 2 above)
Comment (Mark-Simulacrum) [reply src/libcore/hint.rs:84]: Yeah, just leaving feedback in case you're up for making that PR :)

**Comment by Mark-Simulacrum** [comment]
r=me with commits squashed

File: src/libcore/hint.rs (hunk 2 above)
Comment (LeSeulArtichaut) [reply src/libcore/hint.rs:84]: Seems like a low-effort low-hanging fruit. Yum!

**Comment by LeSeulArtichaut** [comment]
@Mark-Simulacrum Rebased! (I can't r+ myself I believe)

**Comment by LeSeulArtichaut** [comment]
I'll try anyway :D
@bors r=Mark-Simulacrum rollup

**Pull request merged**